2.0-proposal/readme.md

```diff
@@ -0,0 +1,183 @@
+
+### Our internal state has changed! [internal]
+
+The internal state of the unit has changed and we need to react to that.
+
+Ex: A file changing (`file.changed.<xyz>`), landscape messing about on the system (removal of `apt.installed.xyz`), a flag being toggled or the data of a flag being changed (`config.changed` unsets `config.checked`).
+
+<!-- Actions is a tricky one. I'd wait a bit to include actions in the story until we're more certain of where we want to go.
+
+### Action requested! [external]
+
+The operator has requested an action that shouldn't significantly impact the state of the application.
+
+Ex: `smoke-test`, `run-job`, `create-backup`, ... -->
+
+# When do triggers and handlers run?
+
+The flow of the framework is the following:
+
+1. The reactive framework will run a handler when all its preconditions are met. A handler that has run is inactive. It will not run again until one of the flags it reacts to has changed. This change can be either the flag being toggled, or the flag data changing.
+
+2. After each handler run, all internal events will be processed. The reactive framework will not remove flags itself. Only triggers and handlers can do that.
+
+<!-- We should do step 2. recursively. This makes it possible for charmers to create deadlocks/ infinite loops. It would be good if the reactive framework could notice an infinite loop and crash. -->
+
+<!-- Triggers or "Connected States" fix a number of issues:  -->
+<!-- multiple handlers can watch the same file for changes by creating a trigger on `file.changed.xyz` that flags `myhandler.file.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/25 https://github.com/juju-solutions/charms.reactive/issues/44 -->
+<!-- when_resource_changed is a trigger that flags `resource.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/87 -->
+
+
+3. After all triggers have been processed, the reactive framework will run the next handler whose preconditions are met. (back to 1.)
+
+The reactive framework goes into sleep mode when there isn't a handler to execute anymore at step 1. The reactive framework wakes up when an event happens.
+
+# What is the difference between external and internal events?
+
+Handlers run transactionally. Internal events are processed after a handler run. External events are processed when the reactive framework wakes up.
+
+ - **All** internal events are processed after **every** handler run.
+ - **Only one** external event is processed when the reactive framework wakes up. If there are still external events waiting to be processed at the moment the reactive framework goes to sleep, it will wake back up immediately and process the next external event.
+
+# Show me the code!
+
+
+```python
+# Register triggers
+charms.reactive.register_trigger(
+  event='config.changed',
+  unset='config.checked')
```

Comment:
> We shouldn't introduce new concepts
> Custom events and triggers seem to add more to the framework


Totally on board with this. I think this warrants an exception because:

1. I see it as the last thing we need to add. It's important that we explore the use-cases before implementing this. We shouldn't add this unless it is the last thing.
2. I see this as "we substitute a bunch of tailor-fit concepts by a generic one". This removes the need for the `when_file_changed` decorator and it removes the need for the special behavior of the `config.changed` flag.
3. Layer devs will be able to use this to create their own `when_file_changed`-like flags. Layer devs continue making smarter layers without needing to add additional stuff to reactive.

> It's very similar to the preflight decorator

That's correct. Our reasoning against preflight is that 1) it is to tied to the inner workings of reactive and hooks and 2) it breaks the concept of "flags define when a handler runs". Having something happen "before any handler runs" is clearly a much needed feature. However, that "something" shouldn't be a handler.

> You can't determine the order that things run in

The order in which handlers run should be determined by the flags and only the flags. That's why we have them. "Layers are snippets of reusable ops code and flags are used for coordination between these layers". This is why I don't want the `trigger` decorator from your comment. We shouldn't add a second way to order handlers. We just need to make the current method a little bit more powerful. The only thing that a trigger can do is change a flag. In that sense they're a lot less powerful than `preflight`. By design. The order of handlers is still coordinated by the flags. We just have gained the ability to connect flags.